### DISCUSSION TITLE
draggable waypoints for geodesic line

### DISCUSSION THREAD

**Discussion by AndrewAnnex** [pyvista/pyvista#2784] [Q&A]
I am interested in a making an interactive "profiling" tool based on multiple draggable "waypoints" in a polyline connected by geodesic paths. A key feature will be the ability to drag a "waypoint" placed on a mesh updating the connecting geodesic paths connected to the given point, similar to how the spline widget allows handles to be dragged. Of course, I don't know ahead of time the positions or number of waypoints a given line could have, and I would like to be able to add new waypoints/delete old waypoints to the line depending on user selection and key combinations. I am mainly interested in the xyz coordinates of the lines/geodesic paths, but I could see this extended to sampling scalars easily also. 

I think a lot of the individual components of this feature are do-able given existing features in pyvista but I am not sure if it is possible to implement the draggable feature component. As I want the point to be on the mesh, the sphere/point must follow the mesh but potentially this could be simplified by avoiding the draggability feature and just inserting a new point into the path after removing the initial point. 

I've placed a few links to relevant features/questions related to this question in case others are curious. 

multiple point picking projected from camera fov:
https://github.com/pyvista/pyvista-support/issues/239

geodesic lines between points:
https://github.com/pyvista/pyvista-support/issues/409 

multiple points:
https://github.com/pyvista/pyvista/discussions/1886 

picking then moving (unanswered):
https://github.com/pyvista/pyvista-support/issues/474

**Comment by AndrewAnnex** [comment]
I recently found the vtkTerrainDataPointPlacer that seems to be exactly what I need, I'm current attempting to use the vtk api directly to try using it https://www.weiy.city/2019/08/use-vtk-to-put-points-on-dem-and-generate-interpolator-line/

**Comment by banesullivan** [comment]
I know I've done this before... It should be a combination of the `enable_geodesic_picking` and `enable_horizon_picking` methods.

I'll see if I can dig up an example

**Reply by AndrewAnnex** [reply]
I also made progress over the weekend directly using the vtk api directly using the vtkContourWidget with both the vtkPolygonalSurfaceContourLineInterpolator and the vtkTerrainContourLineInterpolator methods however I kept crashing python using things incorrectly and the Terrain based methods really required the height data to be in a vtkImageData object. The terrain one did not work as well as it does in the link above however (did not really conform well to the mesh, likely an error on my part though). However getting the interpolated waypoints from the methods did not work as I would hope (ie all of the points interpolated between each node in addition to the nodes).

I would like the method to be somewhat agnostic to the source mesh (uniform grid or polydata) if possible.
